Feature/favorite POIs

tberkane opened the pull request:
Adds the possibility to add pois to favorites by clicking on the star in the forum. The poi will then appear in the favourite pois in the profile

aman3014 commented on app/src/androidTest/java/com/github/epfl/meili/messages/ChatLogMenuButtonsTest.kt:
```diff
@@ -134,15 +134,17 @@ class ChatLogMenuButtonsTest {
-            )
-        ).perform(click())
-    }
+    // This test is not passing on cirrus
+//
+//    @Test
+//    fun clickReviewMenuButton() {
+//
+//        onView(
+//            allOf(
+//                withId(R.id.menu_reviews), withText("Reviews"),
+//            )
+//        ).perform(click())
+//    }
```
This test has been passing on main, it would be weird if it stopped passing. Perhaps it's some other test which fails or causes a time out and this one is shown to be not passing?

MarcelTorne replied:
This has happened to me before. @tberkane did you try running it on your computer with `./gradlew connectedCheck` command? Cause sometimes the tests fail after some time and as @aman3014 has said, the error appears on another one.

aman3014 reviewed (commented):
Could you please define "POIs which were interacted with"? Does it require the user to post to the forum? or chat?

If that's the case, we should discuss with the team if that's really how we want to define the poi history. IMO it would be nice to add a POI to the history as soon as the user was in the vicinity of the POI, as stated in #172.

MarcelTorne commented on app/src/main/java/com/github/epfl/meili/profile/poihistory/PoiHistoryActivity.kt:
```diff
@@ -0,0 +1,96 @@
+
+        fun addPoiToHistory(userKey: String, poi: PointOfInterest) {
+            FirestoreDatabase( // add to poi history
+                "poi-history/$userKey/poi-history",
```
I think we should try to not hardcode the paths and maybe provide a function that will get you the path given the user key. (c.f. the code review done by the assistants

MarcelTorne reviewed (changes requested):
Thank you Thomas for this good job! I just added one nitpick and I am also waiting for your response to @aman3014 's comment on the failing test before approving.

MarcelTorne commented:
> Could you please define "POIs which were interacted with"? Does it require the user to post to the forum? or chat?
> 
> If that's the case, we should discuss with the team if that's really how we want to define the poi history. IMO it would be nice to add a POI to the history as soon as the user was in the vicinity of the POI, as stated in #172 (which should be linked to the PR btw)

What I think would be the best is that you add the POI once the user has the POI reachable (ie he/she is in the vicinity) and the user clicks on its marker.

tberkane commented:
> > Could you please define "POIs which were interacted with"? Does it require the user to post to the forum? or chat?
> > If that's the case, we should discuss with the team if that's really how we want to define the poi history. IMO it would be nice to add a POI to the history as soon as the user was in the vicinity of the POI, as stated in #172 (which should be linked to the PR btw)
> 
> What I think would be the best is that you add the POI once the user has the POI reachable (ie he/she is in the vicinity) and the user clicks on its marker. What do you think @tberkane and @aman3014 ?

Yeah I asked on whatsapp

MarcelTorne reviewed (commented):
I just realised that you are doing what had been already done by me in the PoiMarkerViewModel part. The differences are that you don't have the timestamp there (which could maybe be added) and then there is the functionality of when are we setting it to visited that I thought was what you wanted to change. Maybe you can send me a message and discuss this offline, otherwise check `MapAcitivityViewModel` class and the attributes `mPointsOfInterest` and `mPointsOfInterestStatus`

MarcelTorne commented on app/src/main/java/com/github/epfl/meili/models/VisitedPointOfInterest.kt:
```diff
@@ -0,0 +1,9 @@
+import java.util.*
+
+@Parcelize
+data class VisitedPointOfInterest(var poi: PointOfInterest? = null, var dateVisited: Date? = Calendar.getInstance().time) :
```
@tberkane I am unsure but do you think it would be best to change the name of the class to FavouritePointOfInterest? Maybe not, just wondering.

MarcelTorne commented on app/src/androidTest/java/com/github/epfl/meili/photo/CameraActivityTest.kt:
```diff
@@ -107,7 +107,7 @@ class CameraActivityTest {
         }
 
-        onView(withId(R.id.camera_preview)).perform(touches)
+        //onView(withId(R.id.camera_preview)).perform(touches)
```
Is this line supposed to be commented?

tberkane replied:
Yes for me this test also fails sometimes on main branch, so I thinks it would be better to add a task to fix it.

MarcelTorne reviewed (commented):
Thanks for your work Thomas! Looks really good, just added two questions. But I also wanted to put a general remark that I think would be great to have, would it be possible to have the star button as a toggle? So that it is yellow when it is favourite and black when not. Then if you on it when it's yellow then it becomes black and is also removed from your favourite pois list. What do you think? Maybe we can add it as a task for next sprint

MarcelTorne commented on app/src/main/java/com/github/epfl/meili/profile/favoritepois/FavoritePoisActivity.kt:
```diff
@@ -0,0 +1,97 @@
+    companion object {
+        private const val CARD_PADDING: Int = 30
+        private const val ACTIVITY_TITLE = "Favorite POIs"
+        private const val DB_PATH = "poi-history/%s/poi-history"
```
Sorry, I forgot to mention it in the review but do you think we could change the path to something like `favourite-poi...`?

tberkane commented:
> Thanks for your work Thomas! Looks really good, just added two questions. But I also wanted to put a general remark that I think would be great to have, would it be possible to have the star button as a toggle? So that it is yellow when it is favourite and black when not. Then if you on it when it's yellow then it becomes black and is also removed from your favourite pois list. What do you think? Maybe we can add it as a task for next sprint

Yes good idea

aman3014 commented on app/src/main/java/com/github/epfl/meili/profile/favoritepois/FavoritePoisActivity.kt:
```diff
@@ -0,0 +1,97 @@
+        private const val ACTIVITY_TITLE = "Favorite POIs"
+        private const val DB_PATH = "poi-favorite/%s/poi-favorite"
+
+        fun addPoiToFavorites(userKey: String, poi: PointOfInterest) {
+            FirestoreDatabase( // add to poi favorites
+                String.format(DB_PATH, userKey),
+                FavoritePointOfInterest::class.java
+            ).addElement(poi.uid, FavoritePointOfInterest(poi))
```
This creates a FirestoreDatabase every time a POI is added, which is not great IMO. Instead, the database should be instantiated in the view model of the activity where the database is used (here, Forum).

tberkane replied:
Good catch! I moved the db to the forum's view model. But I did keep this method just for testing purposes.

aman3014 replied:
Thanks for changing this! The method seems redundant to me still, why can't we test the db inside the forum's view model?

aman3014 commented on app/src/main/java/com/github/epfl/meili/profile/favoritepois/FavoritePoisRecyclerAdapter.kt:
```diff
@@ -0,0 +1,40 @@
+
+    class PoiViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        private val poiName: TextView = itemView.findViewById(R.id.poi_name)
+        private val visitedDate: TextView = itemView.findViewById(R.id.visited_date)
```
should we have a visited date for favourite pois? The user could potentially favourite a POI that they haven't visited.

tberkane replied:
Yes i forgot to change this, I converted it to a 'favourited' date. To me it seems like it could be useful to have that.

aman3014 replied:
What could the favourite date be useful for?

aman3014 commented on app/src/main/res/layout/poi.xml:
```diff
@@ -0,0 +1,47 @@
+                android:layout_height="wrap_content"
+                android:textSize="15sp" />
+
+            <TextView
+                android:id="@+id/visited_date"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:textSize="15sp"/>
+
```
same comment about the visited date

aman3014 reviewed (approved):
Thanks for making the changes :)
